// dmdb/src/driver/lob.js
"use strict";

var __extends = this && this.__extends || function() {
    var i = function(t, r) {
        return (i = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, r) {
            t.__proto__ = r;
        } || function(t, r) {
            for (var n in r) r.hasOwnProperty(n) && (t[n] = r[n]);
        })(t, r);
    };
    return function(t, r) {
        function n() {
            this.constructor = t;
        }
        i(t, r), t.prototype = null === r ? Object.create(r) : (n.prototype = r.prototype, 
        new n());
    };
}(), __awaiter = this && this.__awaiter || function(t, u, s, c) {
    return new (s = s || Promise)(function(n, r) {
        function i(t) {
            try {
                o(c.next(t));
            } catch (t) {
                r(t);
            }
        }
        function e(t) {
            try {
                o(c.throw(t));
            } catch (t) {
                r(t);
            }
        }
        function o(t) {
            var r;
            t.done ? n(t.value) : ((r = t.value) instanceof s ? r : new s(function(t) {
                t(r);
            })).then(i, e);
        }
        o((c = c.apply(t, u || [])).next());
    });
}, __generator = this && this.__generator || function(n, i) {
    var e, o, u, s = {
        label: 0,
        sent: function() {
            if (1 & u[0]) throw u[1];
            return u[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: r(0),
        throw: r(1),
        return: r(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function r(r) {
        return function(t) {
            return function(r) {
                if (e) throw new TypeError("Generator is already executing.");
                for (;s; ) try {
                    if (e = 1, o && (u = 2 & r[0] ? o.return : r[0] ? o.throw || ((u = o.return) && u.call(o), 
                    0) : o.next) && !(u = u.call(o, r[1])).done) return u;
                    switch (o = 0, u && (r = [ 2 & r[0], u.value ]), r[0]) {
                      case 0:
                      case 1:
                        u = r;
                        break;

                      case 4:
                        return s.label++, {
                            value: r[1],
                            done: !1
                        };

                      case 5:
                        s.label++, o = r[1], r = [ 0 ];
                        continue;

                      case 7:
                        r = s.ops.pop(), s.trys.pop();
                        continue;

                      default:
                        if (!(u = 0 < (u = s.trys).length && u[u.length - 1]) && (6 === r[0] || 2 === r[0])) {
                            s = 0;
                            continue;
                        }
                        if (3 === r[0] && (!u || r[1] > u[0] && r[1] < u[3])) {
                            s.label = r[1];
                            break;
                        }
                        if (6 === r[0] && s.label < u[1]) {
                            s.label = u[1], u = r;
                            break;
                        }
                        if (u && s.label < u[2]) {
                            s.label = u[2], s.ops.push(r);
                            break;
                        }
                        u[2] && s.ops.pop(), s.trys.pop();
                        continue;
                    }
                    r = i.call(n, s);
                } catch (t) {
                    r = [ 6, t ], o = 0;
                } finally {
                    e = u = 0;
                }
                if (5 & r[0]) throw r[1];
                return {
                    value: r[0] ? r[1] : void 0,
                    done: !0
                };
            }([ r, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var stream_1 = require("stream"), abstractLob_1 = require("../desc/abstractLob"), dm_1 = require("../dm"), error_1 = require("./error"), blob_1 = require("./blob"), clob_1 = require("./clob"), bufferUtil_1 = require("../utils/bufferUtil"), middleware_1 = require("../middleware/middleware"), unenumerable_1 = require("../utils/unenumerable"), Lob = function(i) {
    function e(t, r) {
        var n = i.call(this, r) || this;
        return n.chunkSize = 0, n.pieceSize = 0, n.iLob = t, n.type = t.lobFlag === abstractLob_1.AbstractLob.LOB_FLAG_BYTE ? dm_1.dmdb.BLOB : dm_1.dmdb.CLOB, 
        n.rOffset = 1, n.wOffset = 1, n.newMiddlewares(), unenumerable_1.setUnenumerable(n, [ "middlewares", "nextID" ]), 
        n;
    }
    return __extends(e, i), e.prototype.newMiddlewares = function() {
        this.middlewares = new middleware_1.Middlewares(this.iLob.conn, null, e.nextID++, this.iLob.conn.middlewares);
    }, e.newLob = function(r, n) {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return r !== dm_1.dmdb.BLOB ? [ 3, 2 ] : [ 4, blob_1.Blob.newBlobOfLocal(Buffer.alloc(0), n) ];

                  case 1:
                    return [ 2, new e(t.sent()) ];

                  case 2:
                    return r !== dm_1.dmdb.CLOB ? [ 3, 4 ] : [ 4, clob_1.Clob.newClobOfLocal("", n) ];

                  case 3:
                    return [ 2, new e(t.sent()) ];

                  case 4:
                    throw new Error("Invalid Lob Type");
                }
            });
        });
    }, e.prototype.close = function(r) {
        if (void 0 === r) return this.middlewares.valid ? this.middlewares.reset().lob_close(this) : this.do_close();
        if ("function" != typeof r) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.close().then(function() {
            r(null);
        }).catch(function(t) {
            r(t);
        });
    }, e.prototype.getData = function(r) {
        if (void 0 === r) return this.middlewares.valid ? this.middlewares.reset().lob_getData(this) : this.do_getData();
        if ("function" != typeof r) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.getData().then(function(t) {
            r(null, t);
        }).catch(function(t) {
            r(t, void 0);
        });
    }, e.prototype.getLength = function(r) {
        if (void 0 === r) return this.middlewares.valid ? this.middlewares.reset().lob_getLength(this) : this.do_getLength();
        if ("function" != typeof r) throw error_1.DBError.ECJS_INVALID_PARAM();
        this.getLength().then(function(t) {
            r(null, t);
        }).catch(function(t) {
            r(t);
        });
    }, e.prototype.do_close = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return this.iLob.freed = !0, delete this.iLob.conn, this.iLob.lobFlag, abstractLob_1.AbstractLob.LOB_FLAG_BYTE, 
                delete this.iLob.data, this.emit("close"), [ 2 ];
            });
        });
    }, e.prototype.do_getData = function() {
        return __awaiter(this, void 0, void 0, function() {
            var r;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, this.getLength() ];

                  case 1:
                    return r = t.sent(), this.iLob.lobFlag === abstractLob_1.AbstractLob.LOB_FLAG_BYTE ? [ 2, this.iLob.getBytes(1, r) ] : [ 2, this.iLob.getSubString(1, r) ];
                }
            });
        });
    }, e.prototype.do_getLength = function() {
        return __awaiter(this, void 0, void 0, function() {
            return __generator(this, function(t) {
                return [ 2, this.iLob.getLength() ];
            });
        });
    }, e.prototype._read = function(t) {
        var r = this;
        this.iLob.lobFlag === abstractLob_1.AbstractLob.LOB_FLAG_BYTE ? this.iLob.getBytes(this.rOffset, t).then(function(t) {
            0 !== t.length ? (r.push(t), r.rOffset += t.length) : r.emit("end");
        }).catch(function(t) {
            r.emit("error", t);
        }) : this.iLob.lobFlag === abstractLob_1.AbstractLob.LOB_FLAG_CHAR ? this.iLob.getSubString(this.rOffset, t).then(function(t) {
            0 !== t.length ? (r.push(t), r.rOffset += t.length) : r.emit("end");
        }).catch(function(t) {
            r.emit("error", t);
        }) : this.emit("error", new Error("Fatal error occured in Lob"));
    }, e.prototype._write = function(t, r, n) {
        var i, e, o = this;
        this.iLob.lobFlag === abstractLob_1.AbstractLob.LOB_FLAG_BYTE ? (i = "string" == typeof t ? bufferUtil_1.BufferUtil.b(t, this.iLob.conn.getServerEncoding()) : Buffer.from(t), 
        this.iLob.setBytes(this.wOffset, i, 0, i.length).then(function(t) {
            o.wOffset += t, n(null);
        }).catch(function(t) {
            o.emit("error", t), n(t);
        })) : this.iLob.lobFlag === abstractLob_1.AbstractLob.LOB_FLAG_CHAR ? (e = t instanceof Buffer ? bufferUtil_1.BufferUtil.G(t, this.iLob.conn.getServerEncoding()) : t.toString(), 
        this.iLob.setString(this.wOffset, e, 0, e.length).then(function(t) {
            o.wOffset += t, n(null);
        }).catch(function(t) {
            o.emit("error", t), n(t);
        })) : (this.emit("error", new Error("Fatal error occured in Lob")), n(new Error("Fatal error occured in Lob")));
    }, e.nextID = 0, e;
}(stream_1.Duplex);

exports.Lob = Lob;
