// dmdb/src/desc/abstractLob.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, f, h, u) {
    return new (h = h || Promise)(function(s, i) {
        function r(t) {
            try {
                n(u.next(t));
            } catch (t) {
                i(t);
            }
        }
        function e(t) {
            try {
                n(u.throw(t));
            } catch (t) {
                i(t);
            }
        }
        function n(t) {
            var i;
            t.done ? s(t.value) : ((i = t.value) instanceof h ? i : new h(function(t) {
                t(i);
            })).then(r, e);
        }
        n((u = u.apply(t, f || [])).next());
    });
}, __generator = this && this.__generator || function(s, r) {
    var e, n, f, h = {
        label: 0,
        sent: function() {
            if (1 & f[0]) throw f[1];
            return f[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: i(0),
        throw: i(1),
        return: i(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function i(i) {
        return function(t) {
            return function(i) {
                if (e) throw new TypeError("Generator is already executing.");
                for (;h; ) try {
                    if (e = 1, n && (f = 2 & i[0] ? n.return : i[0] ? n.throw || ((f = n.return) && f.call(n), 
                    0) : n.next) && !(f = f.call(n, i[1])).done) return f;
                    switch (n = 0, f && (i = [ 2 & i[0], f.value ]), i[0]) {
                      case 0:
                      case 1:
                        f = i;
                        break;

                      case 4:
                        return h.label++, {
                            value: i[1],
                            done: !1
                        };

                      case 5:
                        h.label++, n = i[1], i = [ 0 ];
                        continue;

                      case 7:
                        i = h.ops.pop(), h.trys.pop();
                        continue;

                      default:
                        if (!(f = 0 < (f = h.trys).length && f[f.length - 1]) && (6 === i[0] || 2 === i[0])) {
                            h = 0;
                            continue;
                        }
                        if (3 === i[0] && (!f || i[1] > f[0] && i[1] < f[3])) {
                            h.label = i[1];
                            break;
                        }
                        if (6 === i[0] && h.label < f[1]) {
                            h.label = f[1], f = i;
                            break;
                        }
                        if (f && h.label < f[2]) {
                            h.label = f[2], h.ops.push(i);
                            break;
                        }
                        f[2] && h.ops.pop(), h.trys.pop();
                        continue;
                    }
                    i = r.call(s, h);
                } catch (t) {
                    i = [ 6, t ], n = 0;
                } finally {
                    e = f = 0;
                }
                if (5 & i[0]) throw i[1];
                return {
                    value: i[0] ? i[1] : void 0,
                    done: !0
                };
            }([ i, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var const_1 = require("./const"), bufferUtil_1 = require("../utils/bufferUtil"), error_1 = require("../driver/error"), AbstractLob = function() {
    function e(t, i, s, r) {
        this.blobId = BigInt(0), this.inRow = !0, this.groupId = -1, this.fileId = -1, this.pageNo = -1, 
        this.tabId = 0, this.colId = 0, this.rowId = BigInt(0), this.exGroupId = 0, this.exFileId = 0, 
        this.exPageNo = 0, this.curFileId = 0, this.curPageNo = 0, this.curPageOffset = 0, 
        this.totalOffset = 0, this.readOver = !1, this.local = !0, this.updateable = !0, 
        this.length = -1, this.compatibleOracle = !1, this.fetchAll = !1, this.freed = !1, 
        this.modify = !1, this.conn = i, this.lobFlag = t, s && r && (this.compatibleOracle = i.compatibleOracle(), 
        this.local = !1, this.updateable = !r.readonly, this.tabId = r.lobTabId, this.colId = r.lobColId, 
        this.inRow = bufferUtil_1.BufferUtil.e(s, e.NBLOB_HEAD_IN_ROW_FLAG) === e.LOB_IN_ROW, 
        this.blobId = bufferUtil_1.BufferUtil.n(s, e.NBLOB_HEAD_BLOBID), this.inRow || (this.groupId = bufferUtil_1.BufferUtil.r(s, e.NBLOB_HEAD_OUTROW_GROUPID), 
        this.fileId = bufferUtil_1.BufferUtil.r(s, e.NBLOB_HEAD_OUTROW_FILEID), this.pageNo = bufferUtil_1.BufferUtil.t(s, e.NBLOB_HEAD_OUTROW_PAGENO)), 
        i.newLobFlag && (this.tabId = bufferUtil_1.BufferUtil.t(s, e.NBLOB_EX_HEAD_TABLE_ID), 
        this.colId = bufferUtil_1.BufferUtil.r(s, e.NBLOB_EX_HEAD_COL_ID), this.rowId = bufferUtil_1.BufferUtil.n(s, e.NBLOB_EX_HEAD_ROW_ID), 
        this.exGroupId = bufferUtil_1.BufferUtil.r(s, e.NBLOB_EX_HEAD_FPA_GRPID), this.exFileId = bufferUtil_1.BufferUtil.r(s, e.NBLOB_EX_HEAD_FPA_FILEID), 
        this.exPageNo = bufferUtil_1.BufferUtil.t(s, e.NBLOB_EX_HEAD_FPA_PAGENO)), this.resetCurrentInfo());
    }
    return e.prototype.resetCurrentInfo = function() {
        this.curFileId = this.fileId, this.curPageNo = this.pageNo, this.totalOffset = 0, 
        this.curPageOffset = 0;
    }, e.prototype.checkFreed = function() {
        if (this.freed) throw error_1.DBError.ECJS_LOB_FREED();
    }, e.prototype.getLengthFromHead = function(t) {
        return bufferUtil_1.BufferUtil.t(t, e.NBLOB_HEAD_BLOB_LEN);
    }, e.prototype.getLength = function() {
        return __awaiter(this, void 0, void 0, function() {
            var i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return (this.checkFreed(), -1 !== this.length) ? [ 3, 2 ] : (this.conn.checkClosed(), 
                    [ 4, (i = this).conn.access.h(this) ]);

                  case 1:
                    i.length = t.sent(), t.label = 2;

                  case 2:
                    return [ 2, this.length ];
                }
            });
        });
    }, e.prototype.canOptimized = function(t) {
        return !(this.inRow || this.fetchAll || this.local || this.modify || t !== this.conn);
    }, e.prototype.buildCtlData = function() {
        var t = Buffer.alloc(this.conn.newLobFlag ? e.NBLOB_EX_HEAD_SIZE : e.NBLOB_OUTROW_HEAD_SIZE);
        return bufferUtil_1.BufferUtil.l(t, e.NBLOB_HEAD_IN_ROW_FLAG, e.LOB_OFF_ROW), bufferUtil_1.BufferUtil.U(t, e.NBLOB_HEAD_BLOBID, this.blobId), 
        bufferUtil_1.BufferUtil.v(t, e.NBLOB_HEAD_BLOB_LEN, -1), bufferUtil_1.BufferUtil.w(t, e.NBLOB_HEAD_OUTROW_GROUPID, this.groupId), 
        bufferUtil_1.BufferUtil.w(t, e.NBLOB_HEAD_OUTROW_FILEID, this.fileId), bufferUtil_1.BufferUtil.v(t, e.NBLOB_HEAD_OUTROW_PAGENO, this.pageNo), 
        this.conn.newLobFlag && (bufferUtil_1.BufferUtil.v(t, e.NBLOB_EX_HEAD_TABLE_ID, this.tabId), 
        bufferUtil_1.BufferUtil.y(t, e.NBLOB_EX_HEAD_COL_ID, this.colId), bufferUtil_1.BufferUtil.U(t, e.NBLOB_EX_HEAD_ROW_ID, this.rowId), 
        bufferUtil_1.BufferUtil.y(t, e.NBLOB_EX_HEAD_FPA_GRPID, this.exGroupId), bufferUtil_1.BufferUtil.y(t, e.NBLOB_EX_HEAD_FPA_FILEID, this.exFileId), 
        bufferUtil_1.BufferUtil.v(t, e.NBLOB_EX_HEAD_FPA_PAGENO, this.exPageNo)), t;
    }, e.LOB_FLAG_BYTE = 0, e.LOB_FLAG_CHAR = 1, e.LOB_IN_ROW = 1, e.LOB_OFF_ROW = 2, 
    e.NBLOB_HEAD_BLOBID = (e.NBLOB_HEAD_IN_ROW_FLAG = 0) + const_1.Const.BYTE_SIZE, 
    e.NBLOB_HEAD_BLOB_LEN = e.NBLOB_HEAD_BLOBID + const_1.Const.DDWORD_SIZE, e.NBLOB_HEAD_OUTROW_GROUPID = e.NBLOB_HEAD_BLOB_LEN + const_1.Const.ULINT_SIZE, 
    e.NBLOB_HEAD_OUTROW_FILEID = e.NBLOB_HEAD_OUTROW_GROUPID + const_1.Const.USINT_SIZE, 
    e.NBLOB_HEAD_OUTROW_PAGENO = e.NBLOB_HEAD_OUTROW_FILEID + const_1.Const.USINT_SIZE, 
    e.NBLOB_EX_HEAD_TABLE_ID = e.NBLOB_HEAD_OUTROW_PAGENO + const_1.Const.ULINT_SIZE, 
    e.NBLOB_EX_HEAD_COL_ID = e.NBLOB_EX_HEAD_TABLE_ID + const_1.Const.ULINT_SIZE, e.NBLOB_EX_HEAD_ROW_ID = e.NBLOB_EX_HEAD_COL_ID + const_1.Const.USINT_SIZE, 
    e.NBLOB_EX_HEAD_FPA_GRPID = e.NBLOB_EX_HEAD_ROW_ID + const_1.Const.LINT64_SIZE, 
    e.NBLOB_EX_HEAD_FPA_FILEID = e.NBLOB_EX_HEAD_FPA_GRPID + const_1.Const.USINT_SIZE, 
    e.NBLOB_EX_HEAD_FPA_PAGENO = e.NBLOB_EX_HEAD_FPA_FILEID + const_1.Const.USINT_SIZE, 
    e.NBLOB_EX_HEAD_SIZE = e.NBLOB_EX_HEAD_FPA_PAGENO + const_1.Const.ULINT_SIZE, e.NBLOB_OUTROW_HEAD_SIZE = e.NBLOB_HEAD_OUTROW_PAGENO + const_1.Const.ULINT_SIZE, 
    e.NBLOB_INROW_HEAD_SIZE = e.NBLOB_HEAD_BLOB_LEN + const_1.Const.ULINT_SIZE, e;
}();

exports.AbstractLob = AbstractLob;
