// dmdb/src/dm.js
"use strict";

var __awaiter = this && this.__awaiter || function(r, u, c, f) {
    return new (c = c || Promise)(function(e, n) {
        function t(r) {
            try {
                i(f.next(r));
            } catch (r) {
                n(r);
            }
        }
        function o(r) {
            try {
                i(f.throw(r));
            } catch (r) {
                n(r);
            }
        }
        function i(r) {
            var n;
            r.done ? e(r.value) : ((n = r.value) instanceof c ? n : new c(function(r) {
                r(n);
            })).then(t, o);
        }
        i((f = f.apply(r, u || [])).next());
    });
}, __generator = this && this.__generator || function(e, t) {
    var o, i, u, c = {
        label: 0,
        sent: function() {
            if (1 & u[0]) throw u[1];
            return u[1];
        },
        trys: [],
        ops: []
    }, r = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (r[Symbol.iterator] = function() {
        return this;
    }), r;
    function n(n) {
        return function(r) {
            return function(n) {
                if (o) throw new TypeError("Generator is already executing.");
                for (;c; ) try {
                    if (o = 1, i && (u = 2 & n[0] ? i.return : n[0] ? i.throw || ((u = i.return) && u.call(i), 
                    0) : i.next) && !(u = u.call(i, n[1])).done) return u;
                    switch (i = 0, u && (n = [ 2 & n[0], u.value ]), n[0]) {
                      case 0:
                      case 1:
                        u = n;
                        break;

                      case 4:
                        return c.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        c.label++, i = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = c.ops.pop(), c.trys.pop();
                        continue;

                      default:
                        if (!(u = 0 < (u = c.trys).length && u[u.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            c = 0;
                            continue;
                        }
                        if (3 === n[0] && (!u || n[1] > u[0] && n[1] < u[3])) {
                            c.label = n[1];
                            break;
                        }
                        if (6 === n[0] && c.label < u[1]) {
                            c.label = u[1], u = n;
                            break;
                        }
                        if (u && c.label < u[2]) {
                            c.label = u[2], c.ops.push(n);
                            break;
                        }
                        u[2] && c.ops.pop(), c.trys.pop();
                        continue;
                    }
                    n = t.call(e, c);
                } catch (r) {
                    n = [ 6, r ], i = 0;
                } finally {
                    o = u = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, r ]);
        };
    }
}, __read = this && this.__read || function(r, n) {
    var e = "function" == typeof Symbol && r[Symbol.iterator];
    if (!e) return r;
    var t, o, i = e.call(r), u = [];
    try {
        for (;(void 0 === n || 0 < n--) && !(t = i.next()).done; ) u.push(t.value);
    } catch (r) {
        o = {
            error: r
        };
    } finally {
        try {
            t && !t.done && (e = i.return) && e.call(i);
        } finally {
            if (o) throw o.error;
        }
    }
    return u;
}, __spread = this && this.__spread || function() {
    for (var r = [], n = 0; n < arguments.length; n++) r = r.concat(__read(arguments[n]));
    return r;
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var dmdb, connection_1 = require("./driver/connection"), error_1 = require("./driver/error"), properties_1 = require("./desc/properties"), url_1 = require("url"), svcConf_1 = require("./desc/svcConf"), const_1 = require("./desc/const"), lob_1 = require("./driver/lob"), pool_1 = require("./driver/pool"), epGroup_1 = require("./desc/epGroup"), ep_1 = require("./desc/ep");

!function(t) {
    function o(i) {
        return __awaiter(this, void 0, void 0, function() {
            var n, e, t, o;
            return __generator(this, function(r) {
                if (!i.startsWith("dm://")) return [ 2, Promise.reject(new Error("URL must start with 'dm://'")) ];
                for (t in n = url_1.parse(i.trim(), !0), (e = new properties_1.Properties()).set("url", i), 
                n.auth && n.auth.split(":")[0] && (e.set("user", n.auth.split(":")[0]), n.auth.split(":")[1] && e.set("password", n.auth.split(":")[1])), 
                n.query) o = n.query[t], e.set(t, o instanceof Array ? o[o.length - 1] : o);
                return [ 2, u(e, n.host) ];
            });
        });
    }
    function i(o) {
        return __awaiter(this, void 0, void 0, function() {
            var n, e, t;
            return __generator(this, function(r) {
                for (t in o = o || {}, n = new properties_1.Properties(), e = o.connectString || o.connectionString || "", 
                n.setAny("url", e), o) n.setAny(t, o[t]);
                return [ 2, u(n, e) ];
            });
        });
    }
    function u(a, _) {
        return __awaiter(this, void 0, void 0, function() {
            var n, e, t, o, i, u, c, f, s;
            return __generator(this, function(r) {
                switch (r.label) {
                  case 0:
                    return e = a.getTrimString("svcConfPath", ""), [ 4, svcConf_1.SvcConf.load(e, !1) ];

                  case 1:
                    return n = r.sent(), t = a.getTrimString("addressRemap", "") || n.globalProperties.getTrimString("addressRemap", ""), 
                    _ = d(_, t), o = __read([ "", "" ], 2), i = o[0], u = o[1], c = _.lastIndexOf(":"), 
                    u = -1 !== c ? (i = _.substring(0, c), _.substring(c + 1)) : (i = _, const_1.Const.PORT_DEFAULT.toString()), 
                    a.set("host", i), a.set("port", u), f = a.getTrimString("userRemap", "") || n.globalProperties.getTrimString("userRemap", ""), 
                    a.set("user", d(a.getTrimString("user", ""), f)), s = (s = n.serverGroupMap.get(_)) || new epGroup_1.EPGroup(i + ":" + u, [ new ep_1.EP(i, Number.parseInt(u)) ]), 
                    a.setDiffProps(s.props), a.getBoolean("rwSeparate", !1) && (a.setIfNotExist("loginMode", const_1.Const.LOGIN_MODE_PRIMARY_ONLY.toString()), 
                    a.setIfNotExist("loginStatus", const_1.Const.SERVER_STATUS_OPEN.toString()), a.setIfNotExist("doSwitch", "true")), 
                    a.setAny("epGroup", s), a.setDiffProps(n.globalProperties), [ 2, a ];
                }
            });
        });
    }
    function d(r, n) {
        if (!r || !n) return r;
        for (var e = /\(.*?,.*?\)/g, t = null; t = e.exec(n); ) {
            var o = t[0].substring(1, t[0].length - 1).split(",");
            if (o[0].trim() === r) return o[1].trim();
        }
        return r;
    }
    t.EXEC_MANY_FAIL = -1, t.OUT_FORMAT_ARRAY = 4001, t.OUT_FORMAT_OBJECT = 4002, t.BLOB = 2019, 
    t.BUFFER = 2006, t.CLOB = 2017, t.CURSOR = 2021, t.DATE = 2014, t.DEFAULT = 0, t.NUMBER = 2010, 
    t.STRING = 2001, t.BIND_IN = 3001, t.BIND_INOUT = 3002, t.BIND_OUT = 3003, t.POOL_STATUS_OPEN = 6e3, 
    t.POOL_STATUS_DRAINING = 6001, t.POOL_STATUS_CLOSED = 6002, t.STMT_TYPE_UNKNOWN = 0, 
    t.STMT_TYPE_SELECT = 1, t.STMT_TYPE_UPDATE = 2, t.STMT_TYPE_DELETE = 3, t.STMT_TYPE_INSERT = 4, 
    t.STMT_TYPE_CREATE = 5, t.STMT_TYPE_DROP = 6, t.STMT_TYPE_ALTER = 7, t.STMT_TYPE_BEGIN = 8, 
    t.STMT_TYPE_DECLARE = 9, t.STMT_TYPE_CALL = 10, t.STMT_TYPE_EXPLAIN_PLAN = 15, t.STMT_TYPE_MERGE = 16, 
    t.STMT_TYPE_ROLLBACK = 17, t.STMT_TYPE_COMMIT = 21, t.autoCommit = !0, t.extendedMetaData = !1, 
    t.outFormat = t.OUT_FORMAT_ARRAY, t.poolMax = 4, t.poolMin = 0, t.poolTimeout = 60, 
    t.queueMax = 500, t.queueRequests = !0, t.queueTimeout = 6e4, t.fetchAsBuffer = [], 
    t.fetchAsString = [], t.allowDebug = !1, t.pools = new Map(), t.getConnection = function(r, n) {
        if (void 0 === n) {
            var e;
            if (r instanceof Object) return i(r).then(function(r) {
                return e = r, connection_1.Connection.newInstanceOfConnection(e).openConnection(e.getAny("epGroup"));
            }).then(function(r) {
                var n = e.getTrimString("schema", "");
                return "" !== n ? r.execute("set schema " + n).then(function() {
                    return r;
                }) : r;
            });
            if ("string" != typeof r) throw error_1.DBError.ECJS_INVALID_URL_OR_ALIAS_OR_ATTR();
            if (r.startsWith("dm://")) return o(r).then(function(r) {
                return e = r, connection_1.Connection.newInstanceOfConnection(e).openConnection(e.getAny("epGroup"));
            }).then(function(r) {
                var n = e.getTrimString("schema", "");
                return "" !== n ? r.execute("set schema " + n).then(function() {
                    return r;
                }) : r;
            });
            if (!t.pools.has(r)) throw error_1.DBError.ECJS_INVALID_URL_OR_ALIAS_OR_ATTR();
            return t.pools.get(r).getConnection();
        }
        if ("function" != typeof n) throw error_1.DBError.ECJS_INVALID_PARAM();
        t.getConnection(r).then(function(r) {
            n(null, r);
        }).catch(function(r) {
            n(r, void 0);
        });
    }, t.createPool = function(n, e) {
        if (n && void 0 === e) return new pool_1.Pool(n).openPool().then(function(r) {
            if (n.poolAlias) {
                if (t.pools.has(n.poolAlias)) throw error_1.DBError.ECJS_POOL_ALIAS_CONFLICT();
                t.pools.set(n.poolAlias, r);
            }
            return r;
        });
        if (!n || "function" != typeof e) throw error_1.DBError.ECJS_INVALID_PARAM();
        new pool_1.Pool(n).openPool().then(function(r) {
            if (n.poolAlias) {
                if (t.pools.has(n.poolAlias)) throw error_1.DBError.ECJS_POOL_ALIAS_CONFLICT();
                t.pools.set(n.poolAlias, r);
            }
            e(null, r);
        }).catch(function(r) {
            e(r, void 0);
        });
    }, t.getPool = function(r) {
        if (t.pools.has(r)) return t.pools.get(r);
    }, t.isBindParameter = function(r) {
        return null != r && (!("string" == typeof r || "number" == typeof r || r instanceof lob_1.Lob || r instanceof Date || r instanceof BigInt) && (void 0 !== r.dir || void 0 !== r.maxArraySize || void 0 !== r.maxSize || void 0 !== r.type || void 0 !== r.val));
    }, t.parseUrl = o, t.parseConnectionAttr = i, t.debug = function() {
        for (var r = [], n = 0; n < arguments.length; n++) r[n] = arguments[n];
        t.allowDebug && console.log.apply(console, __spread(r));
    };
}(dmdb = exports.dmdb || (exports.dmdb = {}));
