// dmdb/src/net/msg/execDirect.js
"use strict";

var __extends = this && this.__extends || function() {
    var i = function(t, r) {
        return (i = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, r) {
            t.__proto__ = r;
        } || function(t, r) {
            for (var e in r) r.hasOwnProperty(e) && (t[e] = r[e]);
        })(t, r);
    };
    return function(t, r) {
        function e() {
            this.constructor = t;
        }
        i(t, r), t.prototype = null === r ? Object.create(r) : (e.prototype = r.prototype, 
        new e());
    };
}(), __values = this && this.__values || function(t) {
    var r = "function" == typeof Symbol && Symbol.iterator, e = r && t[r], i = 0;
    if (e) return e.call(t);
    if (t && "number" == typeof t.length) return {
        next: function() {
            return t && i >= t.length && (t = void 0), {
                value: t && t[i++],
                done: !t
            };
        }
    };
    throw new TypeError(r ? "Object is not iterable." : "Symbol.iterator is not defined.");
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var execute_1 = require("./execute"), msg_1 = require("./msg"), const_1 = require("../../desc/const"), ExecDirect = function(o) {
    function t(t, r, e, i) {
        var n = o.call(this, t, msg_1.Msg.tr, r, i) || this;
        return n.rr = e, n;
    }
    return __extends(t, o), t.prototype.te = function() {
        var r, t, e, i, n = this.rr.length;
        this.er(n, 1), this.mt.Xn(this.ir.nativeSql, this.ir.connection.getServerEncoding());
        try {
            for (var o = __values(this.rr), s = o.next(); !s.done; s = o.next()) {
                var u = s.value;
                this.mt.In(u.ioType), this.mt.q(u.type), this.mt.q(u.prec), this.mt.q(u.scale);
            }
        } catch (t) {
            r = {
                error: t
            };
        } finally {
            try {
                s && !s.done && (t = o.return) && t.call(o);
            } finally {
                if (r) throw r.error;
            }
        }
        try {
            for (var c = __values(this.rr), f = c.next(); !f.done; f = c.next()) {
                null === (u = f.value).bytes ? this.mt.Tn(const_1.Const.DATA_NULL_LEN) : this.mt.Qn(u.bytes);
            }
        } catch (t) {
            e = {
                error: t
            };
        } finally {
            try {
                f && !f.done && (i = c.return) && i.call(c);
            } finally {
                if (e) throw e.error;
            }
        }
    }, t.prototype.ee = function() {
        return o.prototype.ee.call(this);
    }, t;
}(execute_1.Execute);

exports.ExecDirect = ExecDirect;
