// dmdb/src/net/msg/startup.js
"use strict";

var __extends = this && this.__extends || function() {
    var r = function(t, s) {
        return (r = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, s) {
            t.__proto__ = s;
        } || function(t, s) {
            for (var i in s) s.hasOwnProperty(i) && (t[i] = s[i]);
        })(t, s);
    };
    return function(t, s) {
        function i() {
            this.constructor = t;
        }
        r(t, s), t.prototype = null === s ? Object.create(s) : (i.prototype = s.prototype, 
        new i());
    };
}();

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var msg_1 = require("./msg"), const_1 = require("../../desc/const"), driver_1 = require("../../driver"), error_1 = require("../../driver/error"), driverUtil_1 = require("../../utils/driverUtil"), Startup = function(r) {
    function _(t, s) {
        var i = r.call(this, t, msg_1.Msg.Vt, null, s) || this;
        return i.at = t.at, i;
    }
    return __extends(_, r), _.prototype.te = function() {
        var t, s = this.at.getServerEncoding();
        this.sr = this.at.conn_prop_uKeyName ? 1 : 0, this.mt.v(_.hr, 0), this.mt.v(_.nr, this.at.conn_prop_compress), 
        this.mt.l(_._r, this.at.conn_prop_loginEncrypt ? 1 : 0), this.mt.l(_.or, this.at.conn_prop_loginEncrypt ? 2 : 0), 
        this.mt.l(_.cr, this.at.conn_prop_isBdtaRS ? const_1.Const.RS_BDTA : const_1.Const.RS_ROW), 
        this.mt.l(_.ur, this.at.conn_prop_compressID), this.mt.l(_.vr, "" !== this.at.conn_prop_loginCertificate ? 1 : 0), 
        this.mt.l(_.ar, 0), this.mt.l(_.dr, 1), this.mt.w(_.fr, this.at.msgVersion), this.mt.Vn(driver_1.Driver.VERSION, s), 
        this.mt.In(0), this.sr, this.at.conn_prop_loginEncrypt && (t = this.vs.Nt(), this.mt.Ln(t));
    }, _.prototype.Vs = function() {
        this.mt.Ut(msg_1.Msg.lt);
    }, _.prototype.ee = function() {
        switch (this.at.sslEncrypt = this.mt.t(_.lr), this.mt.t(_.gr), this.mt.t(_.mr)) {
          case 1:
            this.at.serverEncoding = const_1.Const.ENCODING_UTF8;
            break;

          case 2:
            this.at.serverEncoding = const_1.Const.ENCODING_EUCKR;
            break;

          case 3:
            this.at.serverEncoding = const_1.Const.ENCODING_GB18030;
        }
        this.at.conn_prop_compress = this.mt.t(_.Ur);
        var t = this.mt.e(_.br), s = this.mt.e(_.pr);
        this.at.conn_prop_isBdtaRS = 0 < this.mt.e(_.qr), this.at.conn_prop_compressID = this.mt.e(_.xr), 
        this.vs.st = 1 === this.mt.e(_.jr), this.at.newClientType = 1 == this.mt.e(_.wr), 
        this.at.msgVersion = this.mt.oi(_.MSG_RES_STARTUP_PROTOCOL_VERSION), this.$s();
        var i = this.at.getServerEncoding(), r = this.mt.Pn(i);
        if (driverUtil_1.DriverUtil.compareDBVersion(r, driver_1.Driver.VERSION_LOWEST_DATABASE) < 0) throw error_1.DBError.ECJS_ERR_SERVER_VERSION();
        this.at.malini2 = 0 < driverUtil_1.DriverUtil.compareDBVersion(r, driver_1.Driver.VERSION_MALINI2), 
        this.at.execute2 = 0 < driverUtil_1.DriverUtil.compareDBVersion(r, driver_1.Driver.VERSION_EXECUTE2), 
        this.at.lobEmptyCompOrcl = 0 < driverUtil_1.DriverUtil.compareDBVersion(r, driver_1.Driver.VERSION_LOB_EMPTY_COMP_ORCL), 
        this.sr;
        var h = !(this.Ft = -1), n = !1;
        return this.Ot = -1, 0 < s && (this.Ft = this.mt.N()), 0 < t && (-1 === this.Ft ? h = !0 : n = !0, 
        this.jt = this.mt.pn()), 2 === s && (this.Ot = this.mt.N()), this.vs.ot = h, this.vs.ut = n, 
        this;
    }, _.hr = msg_1.Msg.os, _.nr = _.hr + const_1.Const.ULINT_SIZE, _._r = _.nr + const_1.Const.ULINT_SIZE, 
    _.or = _._r + const_1.Const.BYTE_SIZE, _.cr = _.or + const_1.Const.BYTE_SIZE, _.ur = _.cr + const_1.Const.BYTE_SIZE, 
    _.vr = _.ur + const_1.Const.BYTE_SIZE, _.ar = _.vr + const_1.Const.BYTE_SIZE, _.dr = _.ar + const_1.Const.BYTE_SIZE, 
    _.fr = _.dr + const_1.Const.BYTE_SIZE, _.lr = msg_1.Msg.os, _.gr = _.lr + const_1.Const.ULINT_SIZE, 
    _.mr = _.gr + const_1.Const.ULINT_SIZE, _.Ur = _.mr + const_1.Const.ULINT_SIZE, 
    _.Or = _.Ur + const_1.Const.ULINT_SIZE, _.br = _.Or + const_1.Const.ULINT_SIZE, 
    _.pr = _.br + const_1.Const.BYTE_SIZE, _.qr = _.pr + const_1.Const.BYTE_SIZE, _.xr = _.qr + const_1.Const.BYTE_SIZE, 
    _.kr = _.xr + const_1.Const.BYTE_SIZE, _.jr = _.kr + const_1.Const.BYTE_SIZE, _.wr = _.jr + const_1.Const.USINT_SIZE, 
    _.MSG_RES_STARTUP_PROTOCOL_VERSION = _.wr + const_1.Const.BYTE_SIZE, _;
}(msg_1.Msg);

exports.Startup = Startup;
