// dmdb/src/net/access.js
"use strict";

var __extends = this && this.__extends || function() {
    var e = function(t, r) {
        return (e = Object.setPrototypeOf || {
            __proto__: []
        } instanceof Array && function(t, r) {
            t.__proto__ = r;
        } || function(t, r) {
            for (var n in r) r.hasOwnProperty(n) && (t[n] = r[n]);
        })(t, r);
    };
    return function(t, r) {
        function n() {
            this.constructor = t;
        }
        e(t, r), t.prototype = null === r ? Object.create(r) : (n.prototype = r.prototype, 
        new n());
    };
}(), __awaiter = this && this.__awaiter || function(t, u, s, c) {
    return new (s = s || Promise)(function(n, r) {
        function e(t) {
            try {
                o(c.next(t));
            } catch (t) {
                r(t);
            }
        }
        function i(t) {
            try {
                o(c.throw(t));
            } catch (t) {
                r(t);
            }
        }
        function o(t) {
            var r;
            t.done ? n(t.value) : ((r = t.value) instanceof s ? r : new s(function(t) {
                t(r);
            })).then(e, i);
        }
        o((c = c.apply(t, u || [])).next());
    });
}, __generator = this && this.__generator || function(n, e) {
    var i, o, u, s = {
        label: 0,
        sent: function() {
            if (1 & u[0]) throw u[1];
            return u[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: r(0),
        throw: r(1),
        return: r(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function r(r) {
        return function(t) {
            return function(r) {
                if (i) throw new TypeError("Generator is already executing.");
                for (;s; ) try {
                    if (i = 1, o && (u = 2 & r[0] ? o.return : r[0] ? o.throw || ((u = o.return) && u.call(o), 
                    0) : o.next) && !(u = u.call(o, r[1])).done) return u;
                    switch (o = 0, u && (r = [ 2 & r[0], u.value ]), r[0]) {
                      case 0:
                      case 1:
                        u = r;
                        break;

                      case 4:
                        return s.label++, {
                            value: r[1],
                            done: !1
                        };

                      case 5:
                        s.label++, o = r[1], r = [ 0 ];
                        continue;

                      case 7:
                        r = s.ops.pop(), s.trys.pop();
                        continue;

                      default:
                        if (!(u = 0 < (u = s.trys).length && u[u.length - 1]) && (6 === r[0] || 2 === r[0])) {
                            s = 0;
                            continue;
                        }
                        if (3 === r[0] && (!u || r[1] > u[0] && r[1] < u[3])) {
                            s.label = r[1];
                            break;
                        }
                        if (6 === r[0] && s.label < u[1]) {
                            s.label = u[1], u = r;
                            break;
                        }
                        if (u && s.label < u[2]) {
                            s.label = u[2], s.ops.push(r);
                            break;
                        }
                        u[2] && s.ops.pop(), s.trys.pop();
                        continue;
                    }
                    r = e.call(n, s);
                } catch (t) {
                    r = [ 6, t ], o = 0;
                } finally {
                    i = u = 0;
                }
                if (5 & r[0]) throw r[1];
                return {
                    value: r[0] ? r[1] : void 0,
                    done: !0
                };
            }([ r, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var os_1 = require("os"), buffer_1 = require("./buffer"), msg_1 = require("./msg/msg"), startup_1 = require("./msg/startup"), error_1 = require("../driver/error"), const_1 = require("../desc/const"), login_1 = require("./msg/login"), executeRetInfo_1 = require("../desc/executeRetInfo"), parameter_1 = require("../desc/parameter"), stmtAllocate_1 = require("./msg/stmtAllocate"), fetch_1 = require("./msg/fetch"), abstractLob_1 = require("../desc/abstractLob"), enums_1 = require("../desc/enums"), offRowBinder_1 = require("../convert/offRowBinder"), commit_1 = require("./msg/commit"), rollback_1 = require("./msg/rollback"), events_1 = require("events"), msgParser_1 = require("./msgParser"), prepare_1 = require("./msg/prepare"), execute_1 = require("./msg/execute"), getLobData_1 = require("./msg/getLobData"), bufferUtil_1 = require("../utils/bufferUtil"), setLobData_1 = require("./msg/setLobData"), getLobLen_1 = require("./msg/getLobLen"), lobTruncate_1 = require("./msg/lobTruncate"), compressUtil_1 = require("../utils/compressUtil"), preExec_1 = require("./msg/preExec"), bufferArray_1 = require("../utils/bufferArray"), putData_1 = require("./msg/putData"), msgSecurity_1 = require("../security/msgSecurity"), symmCipher_1 = require("../security/symmCipher"), execDirect_1 = require("./msg/execDirect"), stmtFree_1 = require("./msg/stmtFree"), tableTs_1 = require("./msg/tableTs"), moreResult_1 = require("./msg/moreResult"), Access = function(n) {
    function t(t) {
        var r = n.call(this) || this;
        return r.rt = new Array(), r.nt = new msgParser_1.MsgParser({
            onReceive: r.et.bind(r)
        }), r.it = null, r.ot = !1, r.ut = !1, r.st = !1, r.ct = "", r.at = t, r._t = r.at.socket, 
        r;
    }
    return __extends(t, n), t.prototype.ft = function(r) {
        var n = this;
        this.rt.push(r), this.emit("enqueue", r), r.once("error", function(t) {}).once("packet", function(t) {
            r._timer.refresh(), n.emit("data", t);
        }).once("timeout", function() {
            var t = error_1.DBError.ECJS_SOCKET_TIMEOUT();
            t.fatal = !0, n.ht(t, r);
        }), r.once("end", function() {
            r.removeAllListeners(), n.vt(r);
        }), 1 === this.rt.length && this.wt(r);
    }, t.prototype.I = function() {}, t.prototype.R = function(t) {
        return this.nt.write(t), !0;
    }, t.prototype.et = function(t) {
        var r, n, e, i, o, u = this.rt[0];
        u && (u._timer.stop(), u.mt = new buffer_1.DmBuffer(t), u.gt(), 0 < (r = u.dt()) && (this.ut && (n = u.mt.p(msg_1.Msg.lt, r), 
        e = this.bt.decrypt(n, !0), u.mt.pt(msg_1.Msg.lt), u.mt.qt(e), u.yt(e.length)), 
        this.Pt(u) && (r = u.dt(), i = u.mt.p(msg_1.Msg.lt + const_1.Const.ULINT_SIZE, r - const_1.Const.ULINT_SIZE), 
        o = compressUtil_1.CompressUtil.uncompress(i, this.at.conn_prop_compressID), u.mt.pt(msg_1.Msg.lt), 
        u.mt.qt(o), u.yt(o.length))), u.I());
    }, t.prototype.vt = function(t) {
        this.rt.shift();
        var r = this.rt[0];
        r ? this.wt(r) : this.emit("drain");
    }, t.prototype.wt = function(t) {
        try {
            this.at.middlewares.log && this.at.middlewares.log.debug(this.at, this.at.rwStandby ? "accessStandby" : "access", t.xt()), 
            t.At();
            var r, n, e = t.dt();
            if (0 < e && (this.Lt(t) && (r = compressUtil_1.CompressUtil.compress(t.mt.p(msg_1.Msg.lt, e), this.at.conn_prop_compressID), 
            t.mt.pt(msg_1.Msg.lt), t.mt.q(e), t.mt.qt(r), t.yt(const_1.Const.ULINT_SIZE + r.length), 
            t.mt.l(msg_1.Msg.Dt, 1)), this.ut && (e = t.dt(), n = this.bt.encrypt(t.mt.p(msg_1.Msg.lt, e), !0), 
            t.mt.pt(msg_1.Msg.lt), t.mt.qt(n), t.yt(n.length))), t.mt.St > msg_1.Msg.Rt) throw error_1.DBError.ECJS_MSG_TOO_LONG();
            t.kt(), t.mt.Ut(0), 0 < t._timeout && t._timer.start(t._timeout), t.Tt();
        } catch (t) {
            throw this.rt.shift(), t;
        }
    }, t.prototype.Bt = function(t) {
        this.emit("data", t);
    }, t.prototype.P = function() {
        return __awaiter(this, void 0, void 0, function() {
            var r, n;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    return [ 4, this.It() ];

                  case 1:
                    return r = t.sent(), 2 !== this.at.sslEncrypt ? [ 3, 3 ] : [ 4, this.at.startTLS(!1) ];

                  case 2:
                    return t.sent(), [ 3, 5 ];

                  case 3:
                    return 1 !== this.at.sslEncrypt ? [ 3, 5 ] : [ 4, this.at.startTLS(!0) ];

                  case 4:
                    t.sent(), t.label = 5;

                  case 5:
                    return (this.ut || this.ot) && (n = msgSecurity_1.MsgSecurity.computeSessionKey(this.Et, r.jt), 
                    this.Ct(-1 === r.Ft ? msgSecurity_1.MsgSecurity.DES_CFB : r.Ft, n, this.at.conn_prop_cipherPath, -1 === r.Ot ? msgSecurity_1.MsgSecurity.MD5 : r.Ot)), 
                    [ 2, this.Gt() ];
                }
            });
        });
    }, t.prototype.It = function() {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new startup_1.Startup(r, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.Gt = function() {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new login_1.Login(r, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.J = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var e = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, n) {
                    e.ft(new stmtAllocate_1.StmtAllocate(e, i, function(t) {
                        t ? n(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.A = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var e = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, n) {
                    e.ft(new stmtFree_1.StmtFree(e, i, function(t) {
                        t ? n(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.X = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new prepare_1.Prepare(r, i, !1, const_1.Const.EXECUTE_TYPE_ANY, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.Z = function(i, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new execute_1.Execute(r, i, o, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.V = function(i, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new prepare_1.Prepare(r, i, !0, o, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.Q = function(i, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new execDirect_1.ExecDirect(r, i, o, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.Mt = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var e = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, n) {
                    e.ft(new preExec_1.PreExec(e, i, i.params, function(t) {
                        t ? n(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.zt = function(i, o, u, s) {
        return __awaiter(this, void 0, void 0, function() {
            var e = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, n) {
                    e.ft(new putData_1.PutData(e, i, o, u, s, function(t) {
                        t ? n(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.Y = function(u, s, c) {
        return __awaiter(this, void 0, void 0, function() {
            var r, n, e, i, o;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = new Array(u.paramCount), n = 0, t.label = 1;

                  case 1:
                    return n < u.paramCount ? u.params[n].ioType !== parameter_1.Parameter.IO_TYPE_OUT && this.Ht(r, s, n) ? c ? [ 3, 3 ] : [ 4, this.Mt(u) ] : [ 3, 10 ] : [ 3, 11 ];

                  case 2:
                    t.sent(), c = !0, t.label = 3;

                  case 3:
                    return e = s[n], i = new bufferArray_1.BufferArray(), [ 4, e.read(i) ];

                  case 4:
                    t.sent(), o = 0, t.label = 5;

                  case 5:
                    return !e.readOver || 0 < i.length ? !e.readOver && i.length < const_1.Const.MAX_BLOB_LEN_PER_MSG ? [ 4, e.read(i) ] : [ 3, 7 ] : [ 3, 9 ];

                  case 6:
                    t.sent(), t.label = 7;

                  case 7:
                    return o = i.length > const_1.Const.MAX_BLOB_LEN_PER_MSG ? const_1.Const.MAX_BLOB_LEN_PER_MSG : i.length, 
                    [ 4, this.zt(u, n, i, o) ];

                  case 8:
                    return t.sent(), [ 3, 5 ];

                  case 9:
                    r[n] = s[n], t.label = 10;

                  case 10:
                    return n++, [ 3, 1 ];

                  case 11:
                    return [ 2, this.Z(u, [ r ]) ];
                }
            });
        });
    }, t.prototype.tt = function(f, h, v) {
        return __awaiter(this, void 0, void 0, function() {
            var r, n, e, i, o, u, s, c, a, _;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    n = r = 0, e = !1, i = h.length, (o = new executeRetInfo_1.ExecuteRetInfo()).updateCounts = new Array(i).fill(0), 
                    u = new Array(), t.label = 1;

                  case 1:
                    if (!(r < i)) return [ 3, 6 ];
                    for (n = r; n < i; n++) {
                        for (s = h[n], c = new Array(f.paramCount), e = !1, a = 0; a < f.paramCount; a++) if (f.params[a].ioType !== parameter_1.Parameter.IO_TYPE_OUT && this.Ht(c, s, a)) {
                            e = !0;
                            break;
                        }
                        if (e) break;
                        u.push(c);
                    }
                    return n === r ? [ 3, 3 ] : [ 4, this.Z(f, u) ];

                  case 2:
                    _ = t.sent(), u.splice(0), o.union(_, r, n - r), t.label = 3;

                  case 3:
                    return n < i ? [ 4, this.Y(f, h[n], v) ] : [ 3, 5 ];

                  case 4:
                    _ = t.sent(), v = !0, o.union(_, n, 1), t.label = 5;

                  case 5:
                    return r = n + 1, [ 3, 1 ];

                  case 6:
                    return [ 2, o ];
                }
            });
        });
    }, t.prototype.$ = function(s, c, a) {
        return __awaiter(this, void 0, void 0, function() {
            var r, n, e, i, o, u;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = c.length, (n = new executeRetInfo_1.ExecuteRetInfo()).updateCounts = new Array(r), 
                    e = "", i = 0, t.label = 1;

                  case 1:
                    if (!(i < r)) return [ 3, 6 ];
                    t.label = 2;

                  case 2:
                    return t.trys.push([ 2, 4, , 5 ]), [ 4, this.Y(s, c[i], a || 0 !== i) ];

                  case 3:
                    return o = t.sent(), n.union(o, i, 1), [ 3, 5 ];

                  case 4:
                    if (u = t.sent(), n.updateCounts[i] = 0, !s.connection.conn_prop_continueBatchOnError) throw error_1.DBError.ECJS_BATCH_ERROR(u);
                    return e += "row[" + i + "]:" + u + os_1.EOL, [ 3, 5 ];

                  case 5:
                    return i++, [ 3, 1 ];

                  case 6:
                    if (e) throw error_1.DBError.ECJS_BATCH_ERROR(e);
                    return [ 2, n ];
                }
            });
        });
    }, t.prototype.H = function(i, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new fetch_1.Fetch(r, i, BigInt(o), const_1.Const.INT64_MAX, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.T = function() {
        return __awaiter(this, void 0, void 0, function() {
            var e = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, n) {
                    e.ft(new commit_1.Commit(e, function(t) {
                        t ? n(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.D = function() {
        return __awaiter(this, void 0, void 0, function() {
            var e = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(r, n) {
                    e.ft(new rollback_1.Rollback(e, function(t) {
                        t ? n(t) : r();
                    }));
                }) ];
            });
        });
    }, t.prototype.B = function(o, u, s) {
        return __awaiter(this, void 0, void 0, function() {
            var r, n, e, i;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = Buffer.alloc(s), e = n = 0, t.label = 1;

                  case 1:
                    return n < s ? ((e = s - n) > const_1.Const.MAX_BLOB_GET_LEN_PER_MSG && (e = const_1.Const.MAX_BLOB_GET_LEN_PER_MSG), 
                    [ 4, this.Jt(o, u, e) ]) : [ 3, 3 ];

                  case 2:
                    return (i = t.sent()) && 0 !== i.length ? (bufferUtil_1.BufferUtil.L(r, n, i), n += i.length, 
                    u += i.length, o.readOver ? [ 3, 3 ] : [ 3, 1 ]) : [ 3, 3 ];

                  case 3:
                    return [ 2, r ];
                }
            });
        });
    }, t.prototype.C = function(u, s, c) {
        return __awaiter(this, void 0, void 0, function() {
            var r, n, e, i, o;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = "", e = n = 0, t.label = 1;

                  case 1:
                    return n < c ? ((e = c - n) > const_1.Const.MAX_BLOB_GET_LEN_PER_MSG / 2 && (e = const_1.Const.MAX_BLOB_GET_LEN_PER_MSG / 2), 
                    [ 4, this.Jt(u, s, e) ]) : [ 3, 3 ];

                  case 2:
                    return (i = t.sent()) && 0 !== i.length ? (o = bufferUtil_1.BufferUtil.c(i, 0, i.length, u.serverEncoding), 
                    r += o, n += o.length, s += o.length, u.readOver ? [ 3, 3 ] : [ 3, 1 ]) : [ 3, 3 ];

                  case 3:
                    return [ 2, r ];
                }
            });
        });
    }, t.prototype.j = function(_, f, h, v, w) {
        return __awaiter(this, void 0, void 0, function() {
            var r, n, e, i, o, u, s, c, a;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    e = n = r = 0, i = w / const_1.Const.MAX_BLOB_LEN_PER_MSG + 1, u = 1, s = 2, c = o = 0, 
                    t.label = 1;

                  case 1:
                    return c < i ? ((o = 0) === c && (o |= u), c === i - 1 && (o |= s), (e = w - n) > const_1.Const.MAX_BLOB_LEN_PER_MSG && (e = const_1.Const.MAX_BLOB_LEN_PER_MSG), 
                    [ 4, this.Kt(_, o, f, h, v, e) ]) : [ 3, 4 ];

                  case 2:
                    if ((a = t.sent()) <= 0) return [ 3, 4 ];
                    f += a, r += a, n += e, v += e, t.label = 3;

                  case 3:
                    return c++, [ 3, 1 ];

                  case 4:
                    return [ 2, r ];
                }
            });
        });
    }, t.prototype.S = function(v, w, m, g) {
        return __awaiter(this, void 0, void 0, function() {
            var r, n, e, i, o, u, s, c, a, _, f, h;
            return __generator(this, function(t) {
                switch (t.label) {
                  case 0:
                    r = bufferUtil_1.BufferUtil.b(m, g), n = 0, e = r.length, u = o = i = 0, s = e / const_1.Const.MAX_BLOB_LEN_PER_MSG + 1, 
                    a = 1, _ = 2, f = c = 0, t.label = 1;

                  case 1:
                    return f < s ? ((c = 0) == f && (c |= a), f == s - 1 && (c |= _), (u = e - o) > const_1.Const.MAX_BLOB_LEN_PER_MSG && (u = const_1.Const.MAX_BLOB_LEN_PER_MSG), 
                    [ 4, this.Kt(v, c, w, r, n, u) ]) : [ 3, 4 ];

                  case 2:
                    if ((h = t.sent()) <= 0) return [ 3, 4 ];
                    w += h, i += h, o += u, n += u, t.label = 3;

                  case 3:
                    return f++, [ 3, 1 ];

                  case 4:
                    return [ 2, i ];
                }
            });
        });
    }, t.prototype.Jt = function(i, o, u) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new getLobData_1.GetLobData(r, i, o, u, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.Kt = function(i, o, u, s, c, a) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new setLobData_1.SetLobData(r, i, o, u, s, c, a, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.h = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new getLobLen_1.GetLobLen(r, i, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.O = function(i, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new lobTruncate_1.LobTruncate(r, i, o, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.K = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new tableTs_1.TableTS(r, i, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.W = function(i, o) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new moreResult_1.MoreResult(r, i, o, function(t, r) {
                        t ? e(t) : n(r);
                    }));
                }) ];
            });
        });
    }, t.prototype.z = function(i) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, e) {
                    r.ft(new fetch_1.Fetch(r, i, const_1.Const.INT64_MAX, BigInt(1), function(t, r) {
                        t ? e(t) : n(r.updateCount);
                    }));
                }) ];
            });
        });
    }, t.prototype.Nt = function() {
        return this.Et || (this.Et = msgSecurity_1.MsgSecurity.newClientKeyPair()), msgSecurity_1.MsgSecurity.getPublicKey(this.Et);
    }, t.prototype.Ct = function(t, r, n, e) {
        if (0 < t && t < msgSecurity_1.MsgSecurity.MIN_EXTERNAL_CIPHER_ID && r) this.bt = new symmCipher_1.SymmCipher(t, r); else if (t >= msgSecurity_1.MsgSecurity.MIN_EXTERNAL_CIPHER_ID) throw new Error("not supported");
    }, t.prototype.Lt = function(t) {
        return t.Qt !== msg_1.Msg.Vt && this.at.conn_prop_compress !== const_1.Const.MSG_COMPRESS_NO && (this.at.conn_prop_compress === const_1.Const.MSG_COMPRESS_SIMPLE || this.at.conn_prop_compress === const_1.Const.MSG_COMPRESS_AUTO && (!this.at.local && t.dt() > const_1.Const.MSG_COMPRESS_THRESHOLD));
    }, t.prototype.Pt = function(t) {
        return t.Qt !== msg_1.Msg.Vt && this.at.conn_prop_compress !== const_1.Const.MSG_COMPRESS_NO && (this.at.conn_prop_compress === const_1.Const.MSG_COMPRESS_SIMPLE || this.at.conn_prop_compress === const_1.Const.MSG_COMPRESS_AUTO && 1 === t.mt.e(msg_1.Msg.Dt));
    }, t.prototype.Wt = function(t) {
        return t.Qt !== msg_1.Msg.Vt && 1 == this.at.sslEncrypt;
    }, t.prototype.Ht = function(t, r, n) {
        var e, i, o, u = !1;
        return null === r[n] || void 0 === r[n] ? t[n] = enums_1.ParamDataEnum.NULL : r[n] instanceof offRowBinder_1.OffRowBinder ? (u = !0, 
        t[n] = enums_1.ParamDataEnum.OFF_ROW, (e = r[n].obj) instanceof abstractLob_1.AbstractLob && ((i = e).canOptimized(this.at) && (o = i.buildCtlData(), 
        t[n] = new enums_1.LobCtl(o), u = !1))) : t[n] = r[n], u;
    }, t.prototype.F = function(t) {
        t instanceof error_1.DBError && (t.fatal = !0);
        var r = this.rt[0];
        r ? r.I(t) : this.ht(t);
    }, t.prototype.ht = function(r, t) {
        var n = this;
        this.it || r instanceof error_1.DBError && !r.fatal || (this.it = r, process.nextTick(function() {
            n.rt.forEach(function(t) {
                t.I(r);
            }), n.rt.length = 0;
        }), this.emit("end", r));
    }, t;
}(events_1.EventEmitter);

exports.Access = Access;
