// dmdb/src/net/buffer.js
"use strict";

var __awaiter = this && this.__awaiter || function(t, e, s, o) {
    return new (s = s || Promise)(function(i, n) {
        function r(t) {
            try {
                f(o.next(t));
            } catch (t) {
                n(t);
            }
        }
        function u(t) {
            try {
                f(o.throw(t));
            } catch (t) {
                n(t);
            }
        }
        function f(t) {
            var n;
            t.done ? i(t.value) : ((n = t.value) instanceof s ? n : new s(function(t) {
                t(n);
            })).then(r, u);
        }
        f((o = o.apply(t, e || [])).next());
    });
}, __generator = this && this.__generator || function(i, r) {
    var u, f, e, s = {
        label: 0,
        sent: function() {
            if (1 & e[0]) throw e[1];
            return e[1];
        },
        trys: [],
        ops: []
    }, t = {
        next: n(0),
        throw: n(1),
        return: n(2)
    };
    return "function" == typeof Symbol && (t[Symbol.iterator] = function() {
        return this;
    }), t;
    function n(n) {
        return function(t) {
            return function(n) {
                if (u) throw new TypeError("Generator is already executing.");
                for (;s; ) try {
                    if (u = 1, f && (e = 2 & n[0] ? f.return : n[0] ? f.throw || ((e = f.return) && e.call(f), 
                    0) : f.next) && !(e = e.call(f, n[1])).done) return e;
                    switch (f = 0, e && (n = [ 2 & n[0], e.value ]), n[0]) {
                      case 0:
                      case 1:
                        e = n;
                        break;

                      case 4:
                        return s.label++, {
                            value: n[1],
                            done: !1
                        };

                      case 5:
                        s.label++, f = n[1], n = [ 0 ];
                        continue;

                      case 7:
                        n = s.ops.pop(), s.trys.pop();
                        continue;

                      default:
                        if (!(e = 0 < (e = s.trys).length && e[e.length - 1]) && (6 === n[0] || 2 === n[0])) {
                            s = 0;
                            continue;
                        }
                        if (3 === n[0] && (!e || n[1] > e[0] && n[1] < e[3])) {
                            s.label = n[1];
                            break;
                        }
                        if (6 === n[0] && s.label < e[1]) {
                            s.label = e[1], e = n;
                            break;
                        }
                        if (e && s.label < e[2]) {
                            s.label = e[2], s.ops.push(n);
                            break;
                        }
                        e[2] && s.ops.pop(), s.trys.pop();
                        continue;
                    }
                    n = r.call(i, s);
                } catch (t) {
                    n = [ 6, t ], f = 0;
                } finally {
                    u = e = 0;
                }
                if (5 & n[0]) throw n[1];
                return {
                    value: n[0] ? n[1] : void 0,
                    done: !0
                };
            }([ n, t ]);
        };
    }
};

Object.defineProperty(exports, "__esModule", {
    value: !0
});

var bufferUtil_1 = require("../utils/bufferUtil"), msg_1 = require("./msg/msg"), DmBuffer = function() {
    function t(t) {
        this._readOff = 0, this._writeOff = 0, "number" == typeof t ? (t < 0 && (t = 0), 
        this._buf = Buffer.alloc(t)) : (this._buf = Buffer.from(t), this._writeOff = t.length);
    }
    return Object.defineProperty(t.prototype, "Xt", {
        get: function() {
            return this._buf.slice(0, this._writeOff);
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "Yt", {
        get: function() {
            return this._readOff;
        },
        enumerable: !0,
        configurable: !0
    }), Object.defineProperty(t.prototype, "St", {
        get: function() {
            return this._writeOff;
        },
        enumerable: !0,
        configurable: !0
    }), t.prototype.pt = function(t) {
        return this._writeOff = t, this._readOff > t && (this._readOff = t), this;
    }, t.prototype.Ut = function(t) {
        return this._readOff = t, this;
    }, t.prototype.Zt = function() {
        return this.St - this._readOff;
    }, t.prototype.$t = function(t, n, i) {
        return n ? i ? this.tn(t) : this.pt(t) : i ? this._readOff += t : this._readOff -= t, 
        this;
    }, t.prototype.nn = function(f, e) {
        return __awaiter(this, void 0, void 0, function() {
            var u = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(t, n) {
                    setTimeout(function() {
                        n(new Error("Socket read timeout"));
                    }, 3e3);
                    var i = null, r = setInterval(function() {
                        null !== i && (clearInterval(r), u.qt(i), t(i.length));
                    }, 1);
                    f.once("readable", function() {
                        var t = f.read(e), n = t.readInt32LE(msg_1.Msg.in);
                        0 < n && (t = Buffer.concat([ t, f.read(n) ])), i = t;
                    });
                }) ];
            });
        });
    }, t.prototype.rn = function(u) {
        return __awaiter(this, void 0, void 0, function() {
            var r = this;
            return __generator(this, function(t) {
                return [ 2, new Promise(function(n, i) {
                    u.write(r._buf, function(t) {
                        t ? i(t) : (console.log("send msg"), n(r));
                    });
                }) ];
            });
        });
    }, t.prototype.un = function() {
        return bufferUtil_1.BufferUtil.fn(this.en(1));
    }, t.prototype.sn = function() {
        return bufferUtil_1.BufferUtil.on(this.en(1));
    }, t.prototype.hn = function() {
        return bufferUtil_1.BufferUtil.cn(this.en(2));
    }, t.prototype.x = function() {
        return bufferUtil_1.BufferUtil.ln(this.en(2));
    }, t.prototype.N = function() {
        return bufferUtil_1.BufferUtil._n(this.en(4));
    }, t.prototype.bn = function() {
        return bufferUtil_1.BufferUtil.an(this.en(4));
    }, t.prototype.Un = function() {
        return bufferUtil_1.BufferUtil.vn(this.en(8));
    }, t.prototype.mn = function() {
        return bufferUtil_1.BufferUtil.gn(this.en(8));
    }, t.prototype.dn = function() {
        return bufferUtil_1.BufferUtil.wn(this.en(4));
    }, t.prototype.yn = function() {
        return bufferUtil_1.BufferUtil.Bn(this.en(8));
    }, t.prototype.en = function(t) {
        var n = bufferUtil_1.BufferUtil.p(this._buf, this._readOff, t);
        return this._readOff += t, n;
    }, t.prototype.pn = function() {
        return this.en(this.bn());
    }, t.prototype.kn = function() {
        return this.en(this.sn());
    }, t.prototype.On = function() {
        return this.en(this.x());
    }, t.prototype.jn = function() {
        for (var t = 0; 0 != this.sn(); ) t++;
        return this.$t(t, !1, !1), this.en(t);
    }, t.prototype.xn = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this.en(t), n);
    }, t.prototype.Pn = function(t) {
        return bufferUtil_1.BufferUtil.G(this.pn(), t);
    }, t.prototype.Sn = function(t) {
        return bufferUtil_1.BufferUtil.G(this.kn(), t);
    }, t.prototype.g = function(t) {
        return bufferUtil_1.BufferUtil.G(this.On(), t);
    }, t.prototype.qn = function(t) {
        return bufferUtil_1.BufferUtil.G(this.jn(), t);
    }, t.prototype.Dn = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.En(t ? 1 : 0));
    }, t.prototype.In = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.a(t));
    }, t.prototype.Mn = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.En(t));
    }, t.prototype.m = function(t) {
        return this.qt(bufferUtil_1.BufferUtil._(t));
    }, t.prototype.Tn = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.Gn(t));
    }, t.prototype.q = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.s(t));
    }, t.prototype.zn = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.An(t));
    }, t.prototype.Cn = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.o(t));
    }, t.prototype.Fn = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.Hn(t));
    }, t.prototype.Jn = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.f(t));
    }, t.prototype.Kn = function(t) {
        return this.qt(bufferUtil_1.BufferUtil.d(t));
    }, t.prototype.qt = function(t) {
        var n = t.length;
        if (0 == n) return n;
        var i, r, u = this._buf.length;
        return this._writeOff + n > u && (i = Math.max(2 * u, this._writeOff + n), r = Buffer.alloc(i), 
        this._buf = r.fill(this._buf, 0, this._buf.length)), this._buf = this._buf.fill(t, this._writeOff, this._writeOff + n), 
        this._writeOff += n, n;
    }, t.prototype.Ln = function(t) {
        return this.zn(t.length) + this.qt(t);
    }, t.prototype.Nn = function(t) {
        return this.Mn(t.length) + this.qt(t);
    }, t.prototype.Qn = function(t) {
        return this.Tn(t.length) + this.qt(t);
    }, t.prototype.Rn = function(t) {
        return this.qt(t) + this.Mn(0);
    }, t.prototype.Vn = function(t, n) {
        var i = bufferUtil_1.BufferUtil.b(t, n);
        return this.Ln(i);
    }, t.prototype.Wn = function(t, n) {
        var i = bufferUtil_1.BufferUtil.b(t, n);
        return this.Nn(i);
    }, t.prototype.k = function(t, n) {
        var i = bufferUtil_1.BufferUtil.b(t, n);
        return this.Qn(i);
    }, t.prototype.Xn = function(t, n) {
        var i = bufferUtil_1.BufferUtil.b(t, n);
        return this.Rn(i);
    }, t.prototype.l = function(t, n) {
        return this.L(t, bufferUtil_1.BufferUtil.a(n));
    }, t.prototype.Yn = function(t, n) {
        return this.L(t, bufferUtil_1.BufferUtil.En(n));
    }, t.prototype.y = function(t, n) {
        return this.L(t, bufferUtil_1.BufferUtil._(n));
    }, t.prototype.w = function(t, n) {
        return this.L(t, bufferUtil_1.BufferUtil.Gn(n));
    }, t.prototype.v = function(t, n) {
        return this.L(t, bufferUtil_1.BufferUtil.s(n));
    }, t.prototype.Zn = function(t, n) {
        return this.L(t, bufferUtil_1.BufferUtil.An(n));
    }, t.prototype.U = function(t, n) {
        return this.L(t, bufferUtil_1.BufferUtil.o(n));
    }, t.prototype.$n = function(t, n) {
        return this.L(t, bufferUtil_1.BufferUtil.Hn(n));
    }, t.prototype.L = function(t, n) {
        return this._buf.set(n, t), n.length;
    }, t.prototype.ti = function(t, n) {
        return this.Zn(t, n.length) + this.L(t + 4, n);
    }, t.prototype.ni = function(t, n) {
        return this.Yn(t, n.length) + this.L(t + 1, n);
    }, t.prototype.ii = function(t, n) {
        return this.w(t, n.length) + this.L(t + 2, n);
    }, t.prototype.ri = function(t, n) {
        return this.L(t, n) + this.Yn(t + n.length, 0);
    }, t.prototype.ui = function(t, n, i) {
        return this.ti(t, bufferUtil_1.BufferUtil.b(n, i));
    }, t.prototype.fi = function(t, n, i) {
        return this.ni(t, bufferUtil_1.BufferUtil.b(n, i));
    }, t.prototype.ei = function(t, n, i) {
        return this.ii(t, bufferUtil_1.BufferUtil.b(n, i));
    }, t.prototype.si = function(t, n, i) {
        return this.ri(t, bufferUtil_1.BufferUtil.b(n, i));
    }, t.prototype.e = function(t) {
        return bufferUtil_1.BufferUtil.fn(this.p(t, 1));
    }, t.prototype.M = function(t) {
        return bufferUtil_1.BufferUtil.on(this.p(t, 1));
    }, t.prototype.r = function(t) {
        return bufferUtil_1.BufferUtil.cn(this.p(t, 2));
    }, t.prototype.oi = function(t) {
        return bufferUtil_1.BufferUtil.ln(this.p(t, 2));
    }, t.prototype.t = function(t) {
        return bufferUtil_1.BufferUtil._n(this.p(t, 4));
    }, t.prototype.hi = function(t) {
        return bufferUtil_1.BufferUtil.an(this.p(t, 4));
    }, t.prototype.n = function(t) {
        return bufferUtil_1.BufferUtil.vn(this.p(t, 8));
    }, t.prototype.ci = function(t) {
        return bufferUtil_1.BufferUtil.gn(this.p(t, 8));
    }, t.prototype.u = function(t) {
        return bufferUtil_1.BufferUtil.wn(this.p(t, 4));
    }, t.prototype.i = function(t) {
        return bufferUtil_1.BufferUtil.Bn(this.p(t, 8));
    }, t.prototype.p = function(t, n) {
        return this._buf.subarray(t, t + n);
    }, t.prototype.li = function(t) {
        var n = this.hi(t);
        return this.p(t + 4, n);
    }, t.prototype._i = function(t) {
        var n = this.M(t);
        return this.p(t + 1, n);
    }, t.prototype.bi = function(t) {
        var n = this.oi(t);
        return this.p(t + 2, n);
    }, t.prototype.ai = function(t) {
        for (var n = 0; 0 !== this.M(t); ) t++, n++;
        return this.p(t - n, n);
    }, t.prototype.Ui = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this.li(t), n);
    }, t.prototype.vi = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this._i(t), n);
    }, t.prototype.mi = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this.bi(t), n);
    }, t.prototype.gi = function(t, n) {
        return bufferUtil_1.BufferUtil.G(this.ai(t), n);
    }, t.prototype.tn = function(t) {
        var n, i = this._buf.length;
        this._writeOff + t > i && (n = Buffer.alloc(2 * i), this._buf = n.fill(this._buf, 0, i));
        for (var r = 0; r < t; r++) this._buf.writeUInt8(0, this._writeOff++);
        return this;
    }, t;
}();

exports.DmBuffer = DmBuffer;
